### PR TITLE
Create utility that can parse JUnit XML files

### LINKED ISSUE #40: Create utility that can parse JUnit XML files
To get test results data we need to be able to read JUnit XML files from different test runners. For that, we need to be able to build a utility that can parse JUnit XML, support different JUnit XML variations and return data in the object model.

Here are sample JUnit XML files:

[android_fail.xml
](https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/fixtures/ftl_junit_xml/android_fail.xml)
[android_pass.xml
](https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/fixtures/ftl_junit_xml/android_pass.xml)
[ios_pass.xml
](https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/fixtures/ftl_junit_xml/ios_pass.xml)
[Ios_fail.xml
](https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/fixtures/ftl_junit_xml/ios_fail.xml)

Example JUnit parsing test written in Kotlin for the files above
[JUnit parsing test
](https://github.com/Flank/flank/blob/master/test_runner/src/test/kotlin/ftl/reports/xml/JUnitXmlTest.kt)

### PR DESCRIPTION
**Pull request by vladlaba**
Resolves #40 

- [x] Implement JUnitXML related object models
- [x] Implement parsing JUnitXML
- [x] Add documentation
- [x] Cover implementations with tests

### CODE REVIEW

**Comment by bootstraponline** [comment]
Hey, in the future you can create a Draft Pull Request for PRs that are still a work in progress. So far the code looks good.

https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by bootstraponline** [comment]
You can find example XML files to parse in this ticket: 
https://github.com/software-platform/metrics/issues/40

File: guardian/test/utils/junit_xml/element_parser/system_out_parser_test.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import 'package:guardian/utils/junit_xml/junit_xml.dart';
+import 'package:test/test.dart';
+import 'package:xml/xml.dart';
+
+import '../test_utils/xml_string_builder_util.dart';
+import '../test_utils/xml_string_parse_util.dart';
+
+void main() {
+  group('SystemOutParser', () {
+    const text = 'Random text output';
```
Comment (illia-romanenko) [inline guardian/test/utils/junit_xml/element_parser/system_out_parser_test.dart]: Nit: 
```suggestion
    const text = 'Text output';
```
Just to not confuse readers with "Random"

File: guardian/test/utils/junit_xml/element_parser/test_case_error_parser_test.dart [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import 'package:guardian/utils/junit_xml/junit_xml.dart';
+import 'package:test/test.dart';
+import 'package:xml/xml.dart';
+
+import '../test_utils/xml_string_builder_util.dart';
+import '../test_utils/xml_string_parse_util.dart';
+
+void main() {
+  group('TestCaseErrorParser', () {
+    const text = 'Random test case error';
```
Comment (illia-romanenko) [inline guardian/test/utils/junit_xml/element_parser/test_case_error_parser_test.dart]: Nit: 
```suggestion
    const text = 'Test case error';
```
to not confuse reader with "Random"

File: guardian/test/utils/junit_xml/element_parser/test_case_failure_parser_test.dart [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import 'package:guardian/utils/junit_xml/junit_xml.dart';
+import 'package:test/test.dart';
+import 'package:xml/xml.dart';
+
+import '../test_utils/xml_string_builder_util.dart';
+import '../test_utils/xml_string_parse_util.dart';
+
+void main() {
+  group('TestCaseFailureParser', () {
+    const text = 'Random test case failure';
```
Comment (illia-romanenko) [inline guardian/test/utils/junit_xml/element_parser/test_case_failure_parser_test.dart]: Nit: 
```suggestion
    const text = 'Test case failure';
```
to not confuse readers with "Random".

File: guardian/test/utils/junit_xml/element_parser/test_case_skipped_test.dart [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import 'package:guardian/utils/junit_xml/junit_xml.dart';
+import 'package:test/test.dart';
+import 'package:xml/xml.dart';
+
+import '../test_utils/xml_string_builder_util.dart';
+import '../test_utils/xml_string_parse_util.dart';
+
+void main() {
+  group('TestCaseSkippedParser', () {
+    const text = 'Random test case skipped';
```
Comment (illia-romanenko) [inline guardian/test/utils/junit_xml/element_parser/test_case_skipped_test.dart]: Nit:
```suggestion
    const text = 'Test case skipped';
```

File: guardian/test/utils/junit_xml/junit_xml_parser_test.dart [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        ''';
+
+        expect(
+          () => parser.parse(missedTestSuiteAttributesXml),
+          throwsFormatException,
+        );
+      },
+    );
+
+    //<?xml version='1.0' encoding='UTF-8' ?>
```
Comment (illia-romanenko) [inline guardian/test/utils/junit_xml/junit_xml_parser_test.dart]: Nit: remove comment

Comment (illia-romanenko) [reply guardian/test/utils/junit_xml/junit_xml_parser_test.dart]: as probably we can just open exact xml file

Comment (illia-romanenko) [reply guardian/test/utils/junit_xml/junit_xml_parser_test.dart]: comment can be misleading if content of the file changes and this comment not updated

**Pull request merged**